### PR TITLE
SourceStreamMethodResolver refactoring

### PR DESCRIPTION
**Pull request by sungam3r**

### CODE REVIEW

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 1]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
                     var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
+
+                throw new InvalidOperationException("If method returns ValueTask<T> than T must be of type IObservable<> or IAsyncEnumerable<>");
             }
+
             // IObservable<T>
-            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IObservable<>))
+            if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IObservable<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:102]:
```suggestion
            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IObservable<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 2]
```diff
@@ -110,20 +108,19 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
                 }
-                var valueTaskType = typeof(ValueTask<IObservable<object?>>);
-                var constructor = valueTaskType.GetConstructor(new Type[] { typeof(IObservable<object?>) })!;
-                return Compile(Expression.New(constructor, bodyExpression));
+                return Compile(Expression.New(_valueTaskObservableCtor, bodyExpression));
             }
+
             // IAsyncEnumerable<T>
-            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+            if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:115]:
```suggestion
            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 3]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                    {
-                        return Compile(Expression.Call(_castFromValueTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromValueTask2AsyncMethodInfo : _castFromValueTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
                 }
+
                 // ValueTask<IAsyncEnumerable<T>>
-                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:90]:
```suggestion
                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 4]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
                     var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
+
+                throw new InvalidOperationException("If method returns Task<T> than T must be of type IObservable<> or IAsyncEnumerable<>");
             }
+
             // ValueTask<T>
-            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(ValueTask<>))
+            if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(ValueTask<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:77]:
```suggestion
            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(ValueTask<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 5]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                    {
-                        return Compile(Expression.Call(_castFromTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromTask2AsyncMethodInfo : _castFromTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
                 }
+
                 // Task<IAsyncEnumerable<T>>
-                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:65]:
```suggestion
                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 6]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
             {
-                var valueTaskType = typeof(ValueTask<IObservable<object?>>);
-                var constructor = valueTaskType.GetConstructor(new Type[] { typeof(Task<IObservable<object?>>) })!;
-                return Compile(Expression.New(constructor, bodyExpression));
+                return Compile(Expression.New(_valueTaskObservableCtor, bodyExpression));
             }
+
             // Task<T>
-            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(Task<>))
+            if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(Task<>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:52]:
```suggestion
            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(Task<>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 7]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
             {
                 return Compile(bodyExpression);
             }
+
             // Task<IObservable<object?>>
-            else if (bodyExpression.Type == typeof(Task<IObservable<object?>>))
+            if (bodyExpression.Type == typeof(Task<IObservable<object?>>))
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:46]:
```suggestion
            else if (bodyExpression.Type == typeof(Task<IObservable<object?>>))
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 8]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
                 {
                     var innerType = type.GetGenericArguments()[0];
                     var method = _convertFromValueTaskAsyncEnumerableMethodInfo.MakeGenericMethod(innerType);
                     var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
+
+                throw new InvalidOperationException("If method returns ValueTask<T> than T must be of type IObservable<> or IAsyncEnumerable<>");
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:98]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 9]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
                 {
                     var innerType = type.GetGenericArguments()[0];
                     var method = _convertFromTaskAsyncEnumerableMethodInfo.MakeGenericMethod(innerType);
                     var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
+
+                throw new InvalidOperationException("If method returns Task<T> than T must be of type IObservable<> or IAsyncEnumerable<>");
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:73]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 10]
```diff
@@ -110,20 +108,19 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
+            if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
             {
                 var innerType = bodyExpression.Type.GetGenericArguments()[0];
                 var method = _convertFromAsyncEnumerableMethodInfo.MakeGenericMethod(innerType);
                 var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
                 return func(bodyExpression, resolveFieldContextParameter);
             }
 
-            throw new InvalidOperationException("Method must return a IObservable<T>, Task<IObservable<T>> or ValueTask<IObservable<T>> where T is a reference type, or a IAsyncEnumerable<T>, Task<IAsyncEnumerable<T>> or ValueTask<IAsyncEnumerable<T>>.");
+            throw new InvalidOperationException("Method must return a IObservable<T>, Task<IObservable<T>>, ValueTask<IObservable<T>>, IAsyncEnumerable<T>, Task<IAsyncEnumerable<T>> or ValueTask<IAsyncEnumerable<T>>.");
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:123]: This change is not appropriate for this PR.
```suggestion
            throw new InvalidOperationException("Method must return a IObservable<T>, Task<IObservable<T>> or ValueTask<IObservable<T>> where T is a reference type, or a IAsyncEnumerable<T>, Task<IAsyncEnumerable<T>> or ValueTask<IAsyncEnumerable<T>>.");
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 11]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                    if (!innerType.IsValueType)
-                    {
-                        return Compile(Expression.Call(_castFromTaskAsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
-                    else
-                    {
-                        return Compile(Expression.Call(_castFromTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromTask2AsyncMethodInfo : _castFromTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:61]: :+1:

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 12]
```diff
@@ -132,6 +129,8 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
             }
         }
 
+        private static readonly ConstructorInfo _valueTaskObservableCtor = typeof(ValueTask<IObservable<object?>>).GetConstructor(new Type[] { typeof(Task<IObservable<object?>>) })!;
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:132]: :+1:

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 13]
```diff
@@ -41,67 +41,65 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                    if (!innerType.IsValueType)
-                    {
-                        return Compile(Expression.Call(_castFromValueTaskAsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
-                    else
-                    {
-                        return Compile(Expression.Call(_castFromValueTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromValueTask2AsyncMethodInfo : _castFromValueTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:86]: :+1:

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 14]
```diff
@@ -110,20 +108,19 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
                     var ctor = adapterType.GetConstructor(new Type[] { bodyExpression.Type })!;
                     bodyExpression = Expression.New(ctor, bodyExpression);
                 }
-                var valueTaskType = typeof(ValueTask<IObservable<object?>>);
-                var constructor = valueTaskType.GetConstructor(new Type[] { typeof(IObservable<object?>) })!;
-                return Compile(Expression.New(constructor, bodyExpression));
+                return Compile(Expression.New(_valueTaskObservableCtor, bodyExpression));
```
Comment (Shane32) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:111]: :+1:

**Review by Shane32 — CHANGES_REQUESTED** [review]
I would not remove `else`.  I would not add conditional exceptions.  The single exception should show all valid types.

**Comment by Shane32** [comment]
As stated previously, I do not think `else` should be removed.  It does not help clarify the code.  The use of `else` is appropriate here, even if not strictly necessary.

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs (hunk 10 above)
Comment (sungam3r) [reply src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:123]: Why? There is no more that constraint.

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 15]
```diff
@@ -41,30 +41,26 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
             {
                 return Compile(bodyExpression);
             }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:44]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs (hunk 10 above)
Comment (Shane32) [reply src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:119]: There is currently.  PR #3484 has not been merged yet.

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 16]
```diff
@@ -41,30 +41,26 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                var constructor = valueTaskType.GetConstructor(new Type[] { typeof(Task<IObservable<object?>>) })!;
-                return Compile(Expression.New(constructor, bodyExpression));
+                return Compile(Expression.New(_valueTaskObservableCtor, bodyExpression));
             }
+
             // Task<T>
             else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(Task<>))
             {
                 var type = bodyExpression.Type.GetGenericArguments()[0];
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:55]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 17]
```diff
@@ -41,30 +41,26 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                        return Compile(Expression.Call(_castFromTaskAsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
-                    else
-                    {
-                        return Compile(Expression.Call(_castFromTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromTask2AsyncMethodInfo : _castFromTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
                 }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:63]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 18]
```diff
@@ -74,23 +70,20 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
             }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:73]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 19]
```diff
@@ -74,23 +70,20 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
             // ValueTask<T>
             else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(ValueTask<>))
             {
                 var type = bodyExpression.Type.GetGenericArguments()[0];
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:78]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 20]
```diff
@@ -74,23 +70,20 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
...
-                        return Compile(Expression.Call(_castFromValueTaskAsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
-                    else
-                    {
-                        return Compile(Expression.Call(_castFromValueTask2AsyncMethodInfo.MakeGenericMethod(innerType), bodyExpression));
-                    }
+                    var method = innerType.IsValueType ? _castFromValueTask2AsyncMethodInfo : _castFromValueTaskAsyncMethodInfo;
+                    return Compile(Expression.Call(method.MakeGenericMethod(innerType), bodyExpression));
                 }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:86]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs (hunk 10 above)
Comment (Shane32) [reply src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:119]: Oh, if you're targeting my PR, then yes, it makes sense.

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 21]
```diff
@@ -100,6 +93,7 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
                     return func(bodyExpression, resolveFieldContextParameter);
                 }
             }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:96]:
```suggestion
```

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs [hunk 22]
```diff
@@ -110,10 +104,9 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
                     var ctor = adapterType.GetConstructor(new Type[] { bodyExpression.Type })!;
                     bodyExpression = Expression.New(ctor, bodyExpression);
                 }
-                var valueTaskType = typeof(ValueTask<IObservable<object?>>);
-                var constructor = valueTaskType.GetConstructor(new Type[] { typeof(IObservable<object?>) })!;
-                return Compile(Expression.New(constructor, bodyExpression));
+                return Compile(Expression.New(_valueTaskObservableCtor, bodyExpression));
             }
+
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:109]:
```suggestion
```

**Review by Shane32 — APPROVED** [review]

File: src/GraphQL/Resolvers/SourceStreamMethodResolver.cs (hunk 10 above)
Comment (sungam3r) [reply src/GraphQL/Resolvers/SourceStreamMethodResolver.cs:110]: Target is your branch.

**Pull request merged**